### PR TITLE
app-layer-tls: SNI extension support

### PR DESCRIPTION
**Pull request by thus**
Decode TLS client hello handshake to get server name from SNI extension. Add this value to tls-log (extended), output-json-tls (extended), and make it available to lua scripts through TlsGetSNI().

Fixed changes suggested in #1682.

### CODE REVIEW

**Comment by inliniac** [comment]
@jasonish can you review the parsing logic. Need a 2nd opinion :)

**Comment by thus** [comment]
I'll add a detection module for this as mentioned in Feature 1134 (redmine).

File: src/app-layer-ssl.c [hunk 1]
```diff
@@ -143,11 +143,93 @@ static int SSLv3ParseHandshakeType(SSLState *ssl_state, uint8_t *input,
...
+
+            /* skip version */
+            input += 2;
+
+            /* skip random */
+            input += 32;
+
+            /* skip session id */
+            uint8_t session_id_length = *(input++);
+            input += session_id_length;
```
Comment (jasonish) [inline src/app-layer-ssl.c]: Its possible to get here and dereference input with much less than 34 bytes. Check the bounds of input as you progress, and perhaps create a #define for random length, and version length.

File: src/app-layer-ssl.c [hunk 2]
```diff
@@ -143,11 +143,93 @@ static int SSLv3ParseHandshakeType(SSLState *ssl_state, uint8_t *input,
...
+
+            if ((input - initial_input) > input_len)
+                goto end;
+
+            /* skip cipher suites */
+            uint16_t cipher_suites_length = ntohs(*(uint16_t *)input);
+            input += cipher_suites_length + 2;
+
```
Comment (jasonish) [inline src/app-layer-ssl.c]: Similarly here, the bounds are checked at line 157 to make sure there is enough input for the data already processed.  But we should check to make sure there is enough input data left before dereferencing at 161.

**Comment by thus** [comment]
Thanks for the comments, Jason :)

**Pull request closed without merging**